service/interpreter/stateExecutionCounter.go

```diff
@@ -69,40 +69,94 @@ func (e *StateExecutionCounter) CreateNextExecutionId(stateId string) string {
 }
 
 func (e *StateExecutionCounter) MarkStateIdExecutingIfNotYet(stateReqs []StateRequest) error {
+	config := e.configer.Get()
+
 	needsUpdateSA := false
 	numOfNew := 0
 	for _, sr := range stateReqs {
 		if sr.IsResumeRequest() {
 			continue
 		}
 		s := sr.GetStateStartRequest()
-		numOfNew++
-		e.stateIdCurrentlyExecutingCounts[s.StateId]++
-		if e.stateIdCurrentlyExecutingCounts[s.StateId] == 1 {
-			// first time the stateId show up
-			needsUpdateSA = true
+
+		if e.globalVersioner.IsAfterVersionOfExecutingStateIdMode() {
+			switch mode := config.GetExecutingStateIdMode(); mode {
+			case "DISABLED":
+				// do nothing
+			case "ENABLED_FOR_ALL":
+				if e.increaseStateIdCurrentlyExecutingCounts(s) {
+					needsUpdateSA = true
+				}
+			default: // "ENABLED_FOR_STATES_WITH_WAIT_UNTIL" or nil or unrecognized enum value
+				options := s.GetStateOptions()
+				if !options.GetSkipWaitUntil() {
```

Comment:
similar here: `compatibility.GetSkipStartApi(&options)` 

Nit: it would be nice to rename to `compatibility.GetSkipWaitUntilApi(&options)`

Reply:
Done